Searchhead clustering

vidkun opened the pull request:
This implements the ability to configure search head clustering and initiate the captain election process.

Usage is documented in the README and additional information regarding search head clustering can be found in the [Splunk Docs](http://docs.splunk.com/Documentation/Splunk/latest/DistSearch/AboutSHC)

rarsan commented on README.md:
```diff
@@ -138,6 +138,39 @@ clustering in the `setup_clustering` recipe:
+  must be a boolean literal `true` or `false`.
+* `node['splunk']['shclustering']['mode']`: The search head clustering mode of the node within
+  the cluster. This is used to determine if the node needs to bootstrap the shcluster and begin
+  captain election. Must be set using string literal 'member' or 'captain'.
```
Minor comment: "begin captain election" - if you're bootstrapping a node as captain, I don't think the election actually occurs here.

vidkun replied:
Good point. I clarified that (hopefully).

rarsan commented on README.md:
```diff
@@ -138,6 +138,39 @@ clustering in the `setup_clustering` recipe:
+  from other shclusters in the environment. Must be a string. Defaults to `shcluster1`.
+  captain election. Must be set using string literal 'member' or 'captain'.
+* `node['splunk']['shclustering']['replication_factor']`: The replication factor
+  of the shcluster. Defaults to 2, must be a positive integer.
```
For consistency, I propose to make the default replication factor the same as Splunk's default, i.e. 3.

vidkun replied:
Wasn't aware of the default. Updated to match.

rarsan commented on README.md:
```diff
@@ -138,6 +138,39 @@ clustering in the `setup_clustering` recipe:
+  of the shcluster members. Defaults to 9900.
+* `node['splunk']['shclustering']['deployer_url']`: The management url for the
+  shcluster deployer server, must be set to a string such as: `https://deployer.domain.tld:8089`.
+  Defaults to empty.
```
I realize deployer should ideally be pre-configured before cluster setup, but it's not required as part of the initialization. How about making `deployer_url` optional?
Also, how about adding the auto configuration of a deployer node? Perhaps via `['shclustering']['mode']=deployer`...

vidkun replied:
It's now optional. The `server.conf` template now has a conditional to only write out the `conf_deploy_fetch_url` line if the `deployer_url` attribute is not empty.

I planned on adding deployer configuration later. I'm still kicking around ideas for how to best implement that cleanly. Wanted to get the basic shclustering capability in and functioning in the meantime.

rarsan commented on README.md:
```diff
@@ -138,6 +138,39 @@ clustering in the `setup_clustering` recipe:
+  cluster members referenced by their `mgmt_uri`. Currently this will do a Chef search for nodes that
+  are in the same environment, with search head clustering enabled, and with the same
+  cluster label. Alternatively, this can be hard-coded with a list of all shcluster 
+  members including the current node. Must be an array of strings. Defaults to an empty array.
```
I like how you've added support to manually set the list which is useful for environment provisioning based on chef local

rarsan commented on README.md:
```diff
@@ -401,6 +434,44 @@ of clustering stanza in `etc/system/local/server.conf`).
+initiate the captain election process.
+
+In addition to using this recipe for configuring the search head cluster members, you
+will also have to manually configure a stand-alone search head instance to serve as the 
```
Strictly speaking, it doesn't need to be standalone...it could be co-located with a deployment server or a cluster master. As mentioned in another comment, how about supporting this by adding a conditional logic in the same recipe via say a `deployer` mode?

rarsan commented on recipes/setup_shclustering.rb:
```diff
@@ -0,0 +1,99 @@
+
+# unless shcluster members are staticly assigned via the node attribute,
+# try to find the other shcluster members via Chef search
+if shcluster_params['shcluster_members'].empty?
```
Minor comment: for consistency, let's use 2-space indentation.

vidkun replied:
Ha! Nice catch. Fixed that up.

rarsan commented on recipes/setup_shclustering.rb:
```diff
@@ -0,0 +1,99 @@
+
+template "#{shcluster_app_dir}/local/server.conf" do
+	source "shclustering/server.conf.erb"
+	mode 0755
```
I think mode 600 is appropriate permissions here

vidkun replied:
I agree. 600 is much more appropriate. Good catch.

rarsan commented on recipes/setup_shclustering.rb:
```diff
@@ -0,0 +1,99 @@
+    :shcluster_secret => shcluster_secret
+  )
+  sensitive true
+	notifies :restart, 'service[splunk]', :immediately
```
Good call on adding `sensitive` property.
I think this is missing `.setup_shcluster` File.exist check before overriding `server.conf`?
By the way, is it required that Splunk restarts immediately?

vidkun replied:
The immediate restart is for security purposes. Without it, the shcluster_secret is left sitting in the server.conf in plain text. I kicked off the immediate restart so that Splunk will encrypt it.

This restart and the lack of checking for .setup_shcluster should accommodate the ability to make changes to the server.conf on subsequent converges without requiring re-running the init or bootstrap commands.

Does that logic seem sound or did I overlook something?

rarsan replied:
Good call on immediate restart - I'll make the same change in other places setting passphrase. It would also help if chef run fails subsequently.

I think your logic makes sense. After thinking about this, the file checking is not required here as long as subsequent runs (with **same** or **different** attributes) converge successfully.

rarsan commented on recipes/setup_shclustering.rb:
```diff
@@ -0,0 +1,99 @@
+
+# bootstrap the shcluster and elect a captain if initial_captain set to true and this is the initial shcluster build
+if node['splunk']['shclustering']['mode'] == "captain"
+	execute 'bootstrap-shcluster' do
```
May want to move `shcluster_servers_list` generation to here if it's only required under `captain` mode.

vidkun replied:
Agreed. I like that better.

rarsan commented on spec/recipes/setup_shclustering_spec.rb:
```diff
@@ -0,0 +1,138 @@
+	    end
+
+			let(:shcluster_servers_list) do
+				chef_run.node['splunk']['shclustering']['shcluster_members'].join(',')
```
May want to add a new `let` statement for `chef_run` under each context to ensure test independence and new chef run.

vidkun replied:
My test-fu is a bit weak so I'm not sure exactly how that would help out here. Are you suggesting simply duplicating the `let` statement on line 104 to context beginning on line 108?

rarsan replied:
Yes. Do you want to an execute another new chef_run? If so, you'd want re-define `:chef_run` similar to https://github.com/chef-cookbooks/chef-splunk/blob/master/spec/recipes/install_server_spec.rb
or https://github.com/chef-cookbooks/chef-splunk/blob/master/spec/recipes/setup_clustering_spec.rb
I'm not sure but it may be the cause of the error you reported below..

Minor suggestion: may want to manually hardcode `shcluster_servers_list` in the `let` statement to ensure test integrity vs setting it based on node attribute.

vidkun commented:
I think I addressed what you were getting at regarding the tests. The members list is hard coded in the tests now.

Also fixed an error from an extra .join that broke things when setting the members via attributes instead of via chef search.

rarsan commented:
Looks great @vidkun. I pulled your changes locally and made minor edits. Could you give me write access to your repo so you could review? After you approve, I think we're good to go.

rarsan commented:
Finally merged 🎆 
Thank you @vidkun for working on this.
